linalg/invariants.hpp

```diff
@@ -0,0 +1,1347 @@
+      dI2[2] = c1*dI2b[2];
+      dI2[3] = c1*dI2b[3];
+   }
+   void Eval_dI2b()
+   {
+      eval_state |= HAVE_dI2b;
+      // I2b = det(J)
+      // dI2b = adj(J)^T
+      Get_I2b();
+      dI2b[0] =  sign_detJ*J[3];
+      dI2b[1] = -sign_detJ*J[2];
+      dI2b[2] = -sign_detJ*J[1];
+      dI2b[3] =  sign_detJ*J[0];
+   }
+   void Eval_DaJ() // D adj(J) = D dI2b^t
+   {
+      eval_state |= HAVE_DaJ;
+      Get_dI2b();
+      Eval_DZt(dI2b, &DaJ);
+   }
+   void Eval_DJt() // D J^t
+   {
+      eval_state |= HAVE_DJt;
+      Eval_DZt(J, &DJt);
+   }
+   void Eval_DZt(const scalar_t *Z, scalar_t **DZt_ptr)
+   {
+      MFEM_ASSERT(D != NULL, "");
+      const int nd = D_height;
+      scalar_t *DZt = *DZt_ptr;
+      if (DZt == NULL) { *DZt_ptr = DZt = new scalar_t[2*alloc_height]; }
+      for (int i = 0; i < nd; i++)
+      {
+         const int i0 = i+nd*0, i1 = i+nd*1;
+         DZt[i0] = D[i0]*Z[0] + D[i1]*Z[2];
+         DZt[i1] = D[i0]*Z[1] + D[i1]*Z[3];
+      }
+   }
+
+public:
+   /// The Jacobian should use column-major storage.
+   InvariantsEvaluator2D(const scalar_t *Jac = NULL)
+      : J(Jac), D_height(), alloc_height(), D(), DaJ(), DJt(), DXt(), DYt(),
+        eval_state(0) { }
+
+   ~InvariantsEvaluator2D()
+   {
+      delete [] DYt;
+      delete [] DXt;
+      delete [] DJt;
+      delete [] DaJ;
+   }
+
+   /// The Jacobian should use column-major storage.
+   void SetJacobian(const scalar_t *Jac) { J = Jac; eval_state = 0; }
+
+   /// The @a Deriv matrix is `dof x 2`, using column-major storage.
+   void SetDerivativeMatrix(int height, const scalar_t *Deriv)
+   {
+      eval_state &= ~(HAVE_DaJ | HAVE_DJt);
+      if (alloc_height < height)
+      {
+         delete [] DYt; DYt = NULL;
+         delete [] DXt; DXt = NULL;
+         delete [] DJt; DJt = NULL;
+         delete [] DaJ; DaJ = NULL;
+         alloc_height = height;
+      }
+      D_height = height;
+      D = Deriv;
+   }
+
+   scalar_t Get_I1()  { if (dont(HAVE_I1 )) { Eval_I1();  } return I1; }
+   scalar_t Get_I1b() { if (dont(HAVE_I1b)) { Eval_I1b(); } return I1b; }
+   scalar_t Get_I2()  { if (dont(HAVE_I2b)) { Eval_I2b(); } return I2b*I2b; }
+   scalar_t Get_I2b() { if (dont(HAVE_I2b)) { Eval_I2b(); } return I2b; }
+
+   const scalar_t *Get_dI1()
+   {
+      if (dont(HAVE_dI1 )) { Eval_dI1();  } return dI1;
+   }
+   const scalar_t *Get_dI1b()
+   {
+      if (dont(HAVE_dI1b)) { Eval_dI1b(); } return dI1b;
+   }
+   const scalar_t *Get_dI2()
+   {
+      if (dont(HAVE_dI2)) { Eval_dI2(); } return dI2;
+   }
+   const scalar_t *Get_dI2b()
+   {
+      if (dont(HAVE_dI2b)) { Eval_dI2b(); } return dI2b;
+   }
+
+   // Assemble operation for tensor X with components X_jslt:
+   //    A(i+nd*j,k+nd*l) += (\sum_st  w D_is X_jslt D_kt)
+   //    0 <= i,k < nd,  0 <= j,l,s,t < 2
+   // where nd is the height of D, i.e. the number of DOFs in one component.
+
+   void Assemble_ddI1(scalar_t w, scalar_t *A)
+   {
+      // ddI1_jslt = 2 I_jslt = 2 δ_jl δ_st
```

Comment:
Don't think we've had Unicode characters in the code before... Are we sure we want to start now? (cc: @camierjs)

Reply:
There is a string containing `"ΔTE"` in `ex10{,p}.cpp`. Also, in the comments in for class SidreDataCollection, there are some characters that are not standard ascii, not sure if those are unicode or just extended ascii. Anyway, I think we have a precedent already. ;-)